csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,185 @@ class RedundantUseAnalysis : BackwardVisitor {
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}
+  auto getRootDomainsMergedToLeaf = [&](IterDomain* ld) {
+    std::vector<IterDomain*> merged_root_domains;
+    std::vector<int> index_root_domain;
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {root_domain.begin(), root_domain.end()}, {ld});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto outer_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->inner());
+        if (inner_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(inner_iter - root_domain.begin());
+        }
+        if (outer_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*outer_iter);
+          index_root_domain.emplace_back(outer_iter - root_domain.begin());
+        }
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [B2, I1, I3] while the correct order should be [I3, B2, I1]
+    int n_elements = merged_root_domains.size();
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain[a] > index_root_domain[b];
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (int i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted[i] = merged_root_domains[indices[i]];
+    }
+
+    return merged_root_domains_sorted;
+  };
+
+  // Set stride when use extent of merged root domains to index leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. For broadcast
+  // root domain, use the extent of its concretized domain.
+  auto getRootStride = [&](const std::vector<IterDomain*>& merged_root_domains,
+                           const std::unordered_map<IterDomain*, IterDomain*>&
+                               concretized_broadcast_root_domains) {
+    std::vector<Val*> root_stride(merged_root_domains.size());
+    root_stride[0] = IrBuilder::create<Int>(1);
+    for (int i = 1; i < (int)merged_root_domains.size(); i++) {
+      auto pre_crd = merged_root_domains[i - 1];
+      Val* pre_extent = pre_crd->isBroadcast()
+          ? concretized_broadcast_root_domains.at(pre_crd)->extent()
+          : pre_crd->extent();
+      root_stride[i] = IrBuilder::mulExpr(root_stride[i - 1], pre_extent);
+    }
+    return root_stride;
+  };
+
+  // set write stride
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. write_stride_mod
+  // = {len(B3) * len(B2)}. More cases can be found in
```

Comment:
I think this needs to be guaranteed to be a consistent ordered merge like we detect in the contiguous index pass (scheduler/vectorize_helper has this type of analysis as part of it,but so does the logic to understand where we can contiguous index on root domains.

I don't think this analysis would work for:
`[I1, B2, B3]` with merges (in contrast to your example) as `I1*B3*B2`

i.e. merging the outer dims, then the final dim. The predicate could still be valid in this case, but I think it would have to be computed in a different way.

Reply:
***** Note: This is my only blocker.

Reply:
I don't see any good code to steal from vectorize_helper, that seems a bit aggressive compared to what's needed here, but I think you would need to start at roots and make sure the merges are all next to each other and ordered relative to the roots.

Reply:
this can be done by checking `std::distance(outer_iter, inner_iter)`